### PR TITLE
Merge Android changes into master branch

### PR DESCRIPTION
**Pull request by darconeous**
This commit merges the changes introduced for Android Things into the wpantund master branch.

### CODE REVIEW

**Comment by darconeous** [comment]
We can ignore the Google CLA validation failure.

**Comment by darconeous** [comment]
I realize this will take some time to review and digest.

**Comment by darconeous** [comment]
@abtink: ping

File: src/wpantund/NCPInstanceBase.cpp [hunk 1]
```diff
@@ -329,6 +337,13 @@ NCPInstanceBase::property_get_value(
 	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_DaemonEnabled)) {
 		cb(0, boost::any(mEnabled));
 
+	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_NCPRole)) {
+		std::string role = kWPANTUNDRole_Detached;
+		if (mEnabled && ncp_state_is_associated(get_ncp_state())) {
+			role = kWPANTUNDRole_EndDevice;
```
Comment (abtink) [inline src/wpantund/NCPInstanceBase.cpp]: Current code, we seem to return `EndDevice` or `Detached`. Is it intended?
How about getting the `SPINEL_PROP_NET_ROLE` and mapping it to string?

File: src/ncp-spinel/SpinelNCPInstance-Protothreads.cpp [hunk 2]
```diff
@@ -49,36 +49,69 @@ SpinelNCPInstance::vprocess_disabled(int event, va_list args)
...
+				CONTROL_REQUIRE_OUTBOUND_BUFFER_FLUSHED_WITHIN(NCP_DEFAULT_COMMAND_SEND_TIMEOUT, timeout_error);
+				CONTROL_REQUIRE_COMMAND_RESPONSE_WITHIN(NCP_DEFAULT_COMMAND_RESPONSE_TIMEOUT, timeout_error);
+
+				do {
+					EH_WAIT_UNTIL(!IS_EVENT_FROM_NCP(event));
+					EH_WAIT_UNTIL_WITH_TIMEOUT(0.25, IS_EVENT_FROM_NCP(event));
+				} while(!eh_did_timeout);
+
+				// Fake deep sleep.
+				change_ncp_state(DEEP_SLEEP);
```
Comment (abtink) [inline src/ncp-spinel/SpinelNCPInstance-Protothreads.cpp]: How about if we change to `OFFLINE` instead?

File: src/ncp-spinel/SpinelNCPInstance.cpp [hunk 3]
```diff
@@ -978,26 +1036,52 @@ SpinelNCPInstance::property_get_value(
...
 	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_NCPRSSI)) {
-		SIMPLE_SPINEL_GET(SPINEL_PROP_PHY_RSSI, SPINEL_DATATYPE_INT8_S);
+		if (!mEnabled) {
+			cb(kWPANTUNDStatus_InvalidWhenDisabled, boost::any(std::string("Interface Disabled")));
+		} else {
+			SIMPLE_SPINEL_GET(SPINEL_PROP_PHY_RSSI, SPINEL_DATATYPE_INT8_S);
+		}
+
+//	} else if (strcaseequal(key.c_str(), kWPANTUNDProperty_NetworkIsConnected)) {
+//		SIMPLE_SPINEL_GET(SPINEL_PROP_NET_SINGLETON, SPINEL_DATATYPE_BOOL_S);
```
Comment (abtink) [inline src/ncp-spinel/SpinelNCPInstance.cpp]: ? (is this for future?)

File: src/ncp-spinel/SpinelNCPTaskForm.cpp [hunk 4]
```diff
@@ -155,16 +159,37 @@ nl::wpantund::SpinelNCPTaskForm::vprocess_event(int event, va_list args)
...
+
+		// Turn off PROP_NET_IF_UP
+		mNextCommand = SpinelPackData(
+			SPINEL_FRAME_PACK_CMD_PROP_VALUE_SET(SPINEL_DATATYPE_BOOL_S),
+			SPINEL_PROP_NET_IF_UP,
+			false
+		);
+		EH_SPAWN(&mSubPT, vprocess_send_command(event, args));
+		ret = mNextCommandRet;
+		require_noerr(ret, on_error);
```
Comment (abtink) [inline src/ncp-spinel/SpinelNCPTaskForm.cpp]: The new modern behavior seem to allow a `Form` to go through even if we are already part of a network.

We may need to be careful about info from previous network we were part of (e.g., if I remember correctly , OT may not clear network data on IF/Stack down)...

File: src/ncp-spinel/SpinelNCPTaskJoin.cpp [hunk 5]
```diff
@@ -395,7 +419,34 @@ nl::wpantund::SpinelNCPTaskJoin::vprocess_event(int event, va_list args)
...
+	EH_SPAWN(&mSubPT, vprocess_send_command(event, args));
+
+	mInstance->mNetworkKey = Data();
+	mInstance->mNetworkKeyIndex = 0;
+
+	// Issue a Reset
+	mNextCommand = SpinelPackData(SPINEL_FRAME_PACK_CMD_RESET);
+	EH_SPAWN(&mSubPT, vprocess_send_command(event, args));
+
+	finish(mRet);
```
Comment (abtink) [inline src/ncp-spinel/SpinelNCPTaskJoin.cpp]: :+1:

**Comment by abtink** [comment]
@darconeous , I am closing older PRs on wpantund. Please feel free to re-submit if you want. Thanks.

**Pull request closed without merging**